src/metrics-gatherer.ts

```diff
@@ -0,0 +1,230 @@
+		} else {
+			(<prometheus.Counter>this.metrics.counter[name]).inc(labels, val);
+		}
+	}
+
+	// decrement a gauge
+	dec(name : string, 
+		val : number = 1,
+		labels : LabelSet = {}) {
+		// ensure either that this metric already exists, or if not, create a gauge
+		this.ensureExists(name, 'gauge', {
+			labelNames: Object.keys(labels)
+		});
+		if (!this.checkMetricType(name, ['gauge'])) {
+			throw new MetricsGathererError(`Tried to decrement non-gauge metric ${name}`);
+		}
+		(<prometheus.Gauge>this.metrics.gauge[name]).dec(labels, val);
+	}
+
+	// observe a counter metric
+	counter(name : string, 
+		val : number = 1, 
+		labels : LabelSet = {}) {
+		this.ensureExists(name, 'counter', {
+			labelNames: Object.keys(labels)
+		});
+		(<prometheus.Counter>this.metrics.counter[name]).inc(labels, val);
+	}
+
+	// observe a summary metric
+	summary(name : string, 
+		val : number, 
+		labels : LabelSet = {}) {
+		this.ensureExists(name, 'summary', {
+			labelNames: Object.keys(labels)
+		});
+		(<prometheus.Summary>this.metrics.summary[name]).observe(labels, val);
+	}
+
+	// observe a histogram metric
+	histogram(name : string, 
+		val : number, 
+		labels : LabelSet = {}) {
+		this.ensureExists(name, 'histogram', {
+			labelNames: Object.keys(labels)
+		});
+		(<prometheus.Histogram>this.metrics.histogram[name]).observe(labels, val);
+	}
+
+	// observe both a histogram and a summary, adding suffixes to differentiate
+	histogramSummary(name : string,
+		val : number,
+		labels : LabelSet = {}) {
+		this.histogram(`${name}_hist`, val, labels);
+		this.summary(`${name}_summary`, val, labels);		
+	}
+
+	// check that a metric is of the given type(s)
+	checkMetricType(name : string, kinds : string[]) {
+		return kinds.includes(this.kinds[name]);
+	}
+
+	// used declaratively to ensure a given metric of a certain kind exists, 
+	// given some custom params to instantiate it if absent
+	ensureExists(name : string, kind : string, custom : CustomParams = {}) {
+		// if exists, bail early
+		if (this.existMap[name]) {
+			return;
+		}
+		// if already exists with another kind, throw error
+		if (this.kinds[name] && this.kinds[name] != kind) {
+			throw new MetricsGathererError(`tried to use ${name} twice - first as ` +
+					`${this.kinds[name]}, then as ${kind}`);
+		}
+		// if not described, describe (poorly)
+		if (!(name in this.descriptions)) {
+			this.descriptions[name] = `undescribed ${kind} metric`;
+		}
+		// if doesn't exist, create metric
+		if (!this.metrics[kind][name]) {
+			const constructors : ConstructorMap = {
+				'gauge': new MetricConstructor(prometheus.Gauge),
+				'counter': new MetricConstructor(prometheus.Counter),
+				'summary': new MetricConstructor(prometheus.Summary),
+				'histogram': new MetricConstructor(prometheus.Histogram),
+			};
+			// mix provided custom params with custom params given by 
+			// `describe()`
+			custom = Object.assign(custom, this.customParams[name]);
```

Comment:
This can be
```suggestion
			custom = {
				...custom,
				...this.customParams[name]
			};
```
but it'd be even better to just add it to the existing object creation below as
			this.metrics[kind][name] = constructors[kind].create({
				name: name,
				help: this.descriptions[name],
				...custom,
				...this.customParams[name]
			});